blockchain/chain.go

```diff
@@ -310,6 +314,133 @@ type VoteInfo struct {
 	AgendaStatus []ThresholdStateTuple
 }
 
+// SpendJournalPruneFilter represents a prune filter ensuring data
+// needed for recovery processes are retained.
+type SpendJournalPruneFilter struct {
+	db      database.DB
+	mtx     sync.Mutex
+	hashes  map[chainhash.Hash]uint32
+	pruneCh chan *chainhash.Hash
+}
+
+// Ensure rpcPeer implements the rpcserver.Peer interface.
+var _ indexers.SpendJournalFilterer = (*SpendJournalPruneFilter)(nil)
+
+// newSpendJournalPruneFilter initializes a spend journal prune filter.
+func newSpendJournalPruneFilter(ctx context.Context, db database.DB) *SpendJournalPruneFilter {
+	prunefilter := &SpendJournalPruneFilter{
+		db:      db,
+		hashes:  make(map[chainhash.Hash]uint32),
+		pruneCh: make(chan *chainhash.Hash, 12),
+	}
+
+	go prunefilter.handlePruneSignals(ctx)
+
+	return prunefilter
+}
+
+// AddEntry adds the provided hash to the prune filter.
+func (s *SpendJournalPruneFilter) AddEntry(hash *chainhash.Hash) {
+	s.mtx.Lock()
+	s.hashes[*hash]++
+	s.mtx.Unlock()
+}
+
+// RemoveEntry removes the provided hash from the prune filter.
+func (s *SpendJournalPruneFilter) RemoveEntry(hash *chainhash.Hash) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	count := s.hashes[*hash]
+	if count > 1 {
+		s.hashes[*hash]--
+		return
+	}
+
+	delete(s.hashes, *hash)
+	s.pruneCh <- hash
+}
+
+// Exists asserts if the provided hash is part of the prune filter.
+func (s *SpendJournalPruneFilter) Exists(hash *chainhash.Hash) bool {
+	s.mtx.Lock()
+	_, ok := s.hashes[*hash]
+	s.mtx.Unlock()
+	return ok
+}
+
+// handlePruneSignals processes spend journal filter prune signals.
+func (s *SpendJournalPruneFilter) handlePruneSignals(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case hash := <-s.pruneCh:
+			err := s.db.Update(func(dbTx database.Tx) error {
+				return dbRemoveSpendJournalEntry(dbTx, hash)
+			})
+			if err != nil {
+				log.Errorf("spend journal prune error for hash "+
+					"(%s): %v", hash, err)
+			}
+		}
+	}
+}
+
+// prevScript represents script and script version information for a previous
+// outpoint.
+type prevScript struct {
+	scriptVersion uint16
+	pkScript      []byte
+}
+
+// prevScriptsSnapshot represents a snapshot of script and script version
+// information related to previous outpoints from a utxo viewpoint.
+//
+// This implements the indexers.PrevScripter interface.
+type prevScriptsSnapshot struct {
+	entries map[wire.OutPoint]*prevScript
```

Comment:
Following up since there was an offline question about this and I think it'll be good to show why for other reviewers and contributors as well, this shows what I mean:

```Go
package main

import (
	"fmt"
	"runtime"
)

type prevScript struct {
	scriptVersion uint16
	pkScript      []byte
}

func main() {
	var memstats runtime.MemStats
	runtime.ReadMemStats(&memstats)
	mallocs := -memstats.Mallocs
	m := map[string]prevScript{
		"a": prevScript{},
		"b": prevScript{},
		"c": prevScript{},
		"d": prevScript{},
		"e": prevScript{},
		"f": prevScript{},
	}
	runtime.ReadMemStats(&memstats)
	mallocs += memstats.Mallocs
	fmt.Printf("mallocs without pointer: %d\n", mallocs)
	_ = m

	runtime.ReadMemStats(&memstats)
	mallocs = -memstats.Mallocs
	m2 := map[string]*prevScript{
		"a": &prevScript{},
		"b": &prevScript{},
		"c": &prevScript{},
		"d": &prevScript{},
		"e": &prevScript{},
		"f": &prevScript{},
	}
	runtime.ReadMemStats(&memstats)
	mallocs += memstats.Mallocs
	fmt.Printf("mallocs with pointer: %d\n", mallocs)
	_ = m2
}
```

Results:

```
mallocs without pointer: 0
mallocs with pointer: 6
```